### PR TITLE
676 Local ODSP Snapshot Runner App

### PR DESCRIPTION
**Pull request by kian-thompson**
# [AB#676](https://dev.azure.com/fluidframework/internal/_workitems/edit/676)

For more information about how to contribute to this repo, visit [this page](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md).

## Description

This runner app provides the framework for running some functionality on an explicitly provided ODSP snapshot.

### CODE REVIEW

File: tools/local-snapshot-runner-app/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*!
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/.eslintrc.js:1]: I'm not good at naming, but I feel like we can come up with better name for this package.
Mabe ask question in Teams on suggestions?
I'd suggest putting better description in readme.md in this PR, and simply reference to it in such post, i.e. it should be sufficient to describe what it's about not to require more explanations in chat.

As for name, I have this tension - on one hand, implementation is ODSP specific, so maybe name of package should reflect that. On other hand, I can see a world where we will make it more generic (and maybe pluggable) in the future to support other storages. Then notion of "snapshot" should be removed, as it's rather specific to ODSP (we might use other representations, for example fluid-fetch representation that represents each blob as a file)

I'd try to find some generic names that do not put us in the corner in the future.
Maybe "fluid runner"? Does not tell much, but sort of in line with "fluid fetcher" and "replay tool" - you need to know more to tell what it actually does, but once you know, name resonates and sticks.

File: tools/local-snapshot-runner-app/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/.eslintrc.js:3]: BTW, I'd personally put it ino packages/tools. I feel like tools/ should have tools that are mostly around repo management (build tools, keys, bundle size tools, etc.). All the other tools that are more related to runtime are there

Comment (vladsud) [reply tools/local-snapshot-runner-app/.eslintrc.js:3]: Noticed it mostly due to presence of package lock file :)

File: tools/local-snapshot-runner-app/localSnapshotRunnerApp [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env node
+require("./dist/localSnapshotRunnerApp.js");
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/localSnapshotRunnerApp:2]: Should this be in a bin/ folder? See packages\tools\replay-tool\bin, packages\tools\fetch-tool\bin

File: tools/local-snapshot-runner-app/src/packageVersion.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ *
+ * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY
+ */
+
+export const pkgName = "@fluidframework/local-snapshot-runner-app";
+export const pkgVersion = "0.1.0";
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/packageVersion.ts:9]: shouldn't it be 1.2? Does it depend on where code is (physically)?

File: tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot1.json [hunk 5]
```diff
@@ -0,0 +1,148 @@
+{
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot1.json:1]: maybe mention in the name that's ODSP snapshot file?

File: tools/local-snapshot-runner-app/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+export * from "./localSnapshotRunnerApp";
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/index.ts:6]: There is nothing to export from this file.
Instead, it would be great to export exportFile(), such that we can leverage it directly in non-command line environments (for example, in testing, or if someone wants to build a wrapper, or something else).

File: tools/local-snapshot-runner-app/src/exportFile.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        });
+
+        // This needs a ISummaryTree, while what we give to the local ODSP driver is IOdspSnapshot
+        // See LocalOdspDocumentStorageService.getVersions(...) > calls to convertOdspSnapshotToSnapshotTreeAndBlobs
+        // const container = await loader.rehydrateDetachedContainerFromSnapshot(inputFileContent);
+        
+        const container = await loader.resolve({ url: "/fakeUrl/" });
+
+        const results = await codeLoaderBundle.getResults(container, logger);
+        for (const key in results) {
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/exportFile.ts:68]: Any chance you could put lines 34-40, 54-67 (code that works only with in-memory concepts) into separate exported function?

I'd love to be able to use "the core" in pre-production environments with some %% of production flow to start testing it. In such environment, I'll connect it to the write telemetry engine and there would be no files (everything is in memory).
I'll want to provide my own (empty/stock impl) code loader

File: tools/local-snapshot-runner-app/src/exportFile.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    codeLoader: string,
+    inputFile: string,
+    outputFolder: string,
+    scenario: string,
+    telemetryFile: string,
+    loggerOverride?: ITelemetryBaseLogger,
+) {
+    if (fs.existsSync(telemetryFile)) {
+        console.error("Telemetry file already exists. " + telemetryFile);
+        throw new Error("Telemetry file already exists.");
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/exportFile.ts:27]: I'd suggest having a delegate that does lines 26-27.
- For stand-alone tool, it should print to console and invoke process.exit(1) - users do not need to see long stack for where issue happened.
    - That said, even for stand-alone tool we may choose to use it (morph it to take different form) to process many files in a row - in such case process.exit() is not the right tool.
- But if we leverage the "core" of the solution somewhere else (see my other comment), I'd want to throw exception and not print anything to console.
- if I host it in multi-threaded node.js process, I may want to communicate error to main thread and exit thread (see how it's done in snapshot tests).

Basically, I'd love to see code reusable, and only very small piece of code assume that it's stand-alone node.js executable.

Comment (vladsud) [reply tools/local-snapshot-runner-app/src/exportFile.ts:27]: Or maybe even better - not do it through delegate, but through strongly typed return type - it's either success or error of some sort. And outer code controls how this error is handled.
We better not rely on exceptions for that - we regret that same choice in drivers - exceptions should be used for exceptional cases (like logical errors in our logic, out of memory, etc.). Everything else - structured (typed) return types.

In this case (as opposed to a driver example) one type of error is likely sufficient.

File: tools/local-snapshot-runner-app/src/codeLoaderBundle.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+     * Get the scope object to provide at Loader creation
+     */
+    getLoaderScope(): Promise<FluidObject | undefined>;
+
+    /**
+     * Get the results to write from the provided bundle
+     * @param container - container created by this application
+     * @param logger
+     */
+    getResults(container: IContainer, logger: ITelemetryBaseLogger): Promise<Record<string, string>>;
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/codeLoaderBundle.ts:30]: I'd call it something like execute() or run(). getResults assumes not-too expensive type of code

File: tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json [hunk 10]
```diff
@@ -0,0 +1 @@
..."7cfd6ca4-c526-4cf3-9ae1-135e04727e59","clientSequenceNumber":4,"contents":"null","minimumSequenceNumber":69,"referenceSequenceNumber":70,"sequenceNumber":71,"term":1,"timestamp":1652216640394,"type":"noop"}},{"sequenceNumber":72,"op":{"clientId":"5d281acb-967d-45e4-ae57-2a7734e38fe4","clientSequenceNumber":58,"contents":"null","minimumSequenceNumber":70,"referenceSequenceNumber":71,"sequenceNumber":72,"term":1,"timestamp":1652216640861,"type":"noop"}},{"sequenceNumber":73,"op":{"clientId":null,"clientSequenceNumber":-1,"contents":null,"minimumSequenceNumber":70,"referenceSequenceNumber":-1,"sequenceNumber":73,"term":1,"timestamp":1652216642694,"type":"leave","data":"\"5d281acb-967d-45e4-ae57-2a7734e38fe4\""}},{"sequenceNumber":74,"op":{"clientId":null,"clientSequenceNumber":-1,"contents":null,"minimumSequenceNumber":74,"referenceSequenceNumber":-1,"sequenceNumber":74,"term":1,"timestamp":1652216642698,"type":"leave","data":"\"7cfd6ca4-c526-4cf3-9ae1-135e04727e59\""}},{"sequenceNumber":75,"op":{"clientId":null,"clientSequenceNumber":-1,"contents":null,"minimumSequenceNumber":75,"referenceSequenceNumber":75,"sequenceNumber":75,"term":1,"timestamp":1652216642760,"type":"noClient"}}]}
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json:1]: Your other file is nicely formatted for human readability - it would be great for this one to be formatted that way too

File: tools/local-snapshot-runner-app/.eslintrc.js (hunk 2 above)
Comment (anthony-murphy) [reply tools/local-snapshot-runner-app/.eslintrc.js:3]: +1 packages/tools is the right place for something like this, as we want it to align with our client packages

File: tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json (hunk 10 above)
Comment (anthony-murphy) [reply tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json:1]: can we leverage the test data here?

Comment (anthony-murphy) [reply tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json:1]: https://github.com/microsoft/FluidFrameworkTestData

File: tools/local-snapshot-runner-app/src/logger/FileLogger.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+     * Appending each line to file right away for now
+     * @param event - TODO
+     */
+    public send(event: ITelemetryBaseEvent): void {
+        const logEvent = JSON.stringify(event);
+        console.log(logEvent);
+
+        fs.appendFileSync(this.fileName, `${logEvent}\n`);
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/logger/FileLogger.ts:27]: That might be terribly slow - appends on most OSes is not fast. FileLogger class in service tests assumes accumulation and flushing on some cadence / end of test. I like it better, though there is a risk of losing data.
I'd say two things:
- As long as we know how to collect data to learn it's a perf problem (i.e. feedback mechanism), then we can leave it as is and improve if data suggests us so
- The most interesting data that I'd love not to lose is errors. So maybe we can start with a place where it's mostly performant (accumulates logs in memory), but flushes on any category="error" or end of run.

File: tools/local-snapshot-runner-app/src/exportFile.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    const logger = ChildLogger.create(loggerOverride ?? new FileLogger(telemetryFile), "LocalSnapshotRunnerApp");
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/exportFile.ts:30]: not sure about possible override (not sure I see a scenario), but for file logger we want to add a property (functor) that outputs current time. With other workflows, we rely on external implementation of telemetry interface to add a bunch of common properties include event sequence, event time, some correlation IDs, machine info, etc. In general, most of that should be not responsibility of this code (external code can do it), but not for time - once the event is written and time is not captured, there is no way to fill that data.

File: tools/local-snapshot-runner-app/src/fakeUrlResolver.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Fake URL resolver that returns hard coded values on every request
+ */
+export class FakeUrlResolver implements IUrlResolver {
+
+    public async resolve(request: IRequest): Promise<IResolvedUrl | undefined> {
+        const fakeOdspResolvedUrl: IOdspResolvedUrl = {
+            type: "fluid",
+            odspResolvedUrl: true,
+            id: "1",
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:19]: I'd use some unique strings instead of "1". Some of these will get to telemetry stream and someone will wonder where they are coming from. It's so much easier to find unique strings...

File: tools/local-snapshot-runner-app/src/fakeUrlResolver.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            siteUrl: request.url,
+            driveId: "1",
+            itemId: "1",
+            url: request.url,
+            hashedDocumentId: "1",
+            endpoints: {
+                snapshotStorageUrl: request.url,
```
Comment (vladsud) [inline tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:26]: Can we put undefined for URIs (even it TS time does not allow it)? I'd rather see code crashing on undefined then do network call (assuming we made a mistake and got to a place that is about to try to reach out to network). It feels like it would be easier to debug (having callstack)

File: tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json (hunk 10 above)
Comment (kian-thompson) [reply tools/local-snapshot-runner-app/src/test/localSnapshots/localSnapshot2.json:1]: From what I can tell, those snapshots are not ODSP snapshots. I remember looking into this when I worked on the local ODSP driver, and it was decided to keep the snapshot files in the tests.

File: tools/local-snapshot-runner-app/src/test/sampleCodeLoader.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        load: async (): Promise<IFluidModuleWithDetails> => {
+            return {
+                module: { fluidExport: new BaseContainerRuntimeFactory(new Map()) },
+                details: { package: "no-dynamic-package", config: {} },
+            };
+        },
+    };
+}
+
+export async function getLoaderScope(): Promise<FluidObject | undefined> {
```
[the remark this answers is not in the document]
Comment (kian-thompson) [reply tools/local-snapshot-runner-app/src/test/sampleCodeLoader.ts:22]: @vladsud What is the purpose in wrapping the `IFluidFileConverter` in `FluidObject`?

File: tools/local-snapshot-runner-app/src/fakeUrlResolver.ts (hunk 14 above)
Comment (kian-thompson) [reply tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:26]: Do you mean to use something like `undefined!`?

File: tools/local-snapshot-runner-app/src/test/sampleCodeLoader.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (vladsud) [reply tools/local-snapshot-runner-app/src/test/sampleCodeLoader.ts:22]: Probably none? :)

File: tools/local-snapshot-runner-app/src/fakeUrlResolver.ts (hunk 14 above)
Comment (vladsud) [reply tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:26]: undefined is good if code allows it (does not assert it has to be defined) - that seems like the best outcome.
If we have to use some values (for example, driverId & itemId will be used to generate docId that is added to all telemetry), then I'd prefer to see some unique value that is easy to find in code base. Guid will do it, but they are too long and carry less context (without search repo). "LocalOdspDriver" seems like reasonable string to use (or something like that)

Comment (kian-thompson) [reply tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:26]: `IOdspResolvedUrl` requires it to be defined. Would something like `FluidRunner-FakeUrlResolver` work? Or should I reuse the guid used for all the other string properties?

Comment (vladsud) [reply tools/local-snapshot-runner-app/src/fakeUrlResolver.ts:26]: I'd just use "FakeUrlResolver". It's sufficient, and code can move (over time) and be reused, so I'd not couple it to particular usage (FluidRunner)

File: packages/tools/fluid-runner/src/exportFile.ts [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+                    eventName: "Client_ArgsValidationError",
+                    message: argsValidationError,
+                });
+                return { success: false, errorMessage: argsValidationError };
+            }
+
+            // TODO: read file stream
+            const inputFileContent = fs.readFileSync(inputFile, { encoding: "utf-8" });
+
+            const results = await createContainerAndExecute(inputFileContent, logger, await codeLoaderBundle.fluidExport);
```
Comment (ke-at-microsoft) [inline packages/tools/fluid-runner/src/exportFile.ts:68]: @leeviana I was chatting with Teja earlier about this code. If I understand correctly, this is where office-web-host may fit in? Here in this example, "inputFileContent" seems to be a string that's a snapshot. And "codeLoaderBundle.fluidExport" seems to be a construct to help us to load the proper container. I wonder if you can shed some light on how office-web-host may fit here (set aside the problem of office-web-host is not published to a public feed so this program can't access it directly).

Comment (vladsud) [reply packages/tools/fluid-runner/src/exportFile.ts]: Here is my naive understanding. There are couple layers to it, and you can integrate at any layer, in priority order (i.e. the lower on the list, the less supported / ideal is the flow):
1. We give you IContainer. Best, because we control how all the pieces are tied together (URI resolver, driver, etc.).
2. We given you ILoader. Worse, because you need to know what request to use to create container (we should assume that the system is setup in a way where it creates container for only one file instance, and that instance is hard coded via specific URI, asking for any other URI to be resolved will produce an error).
3. We give you URI resolver, driver factory, URI to use, file content (buffer), maybe loader options, maybe something else, and you are responsible for tying it all together. Obviously, there are a lot of inner details that are exposed here which will make it extremely hard to change anything if we need to. I.e. if we need to switch to different workflow (let's say use different driver instance), then it will be more expensive to do so across whole stack.

@leeviana - would love to hear what are your thoughts here.

Comment (leeviana) [reply packages/tools/fluid-runner/src/exportFile.ts]: Yes, let me also copy over my code snippet/response from the Teams chat here. Here's my suggested integration given this code:

Comment (leeviana) [reply packages/tools/fluid-runner/src/exportFile.ts]: const hostConfig = { logger }
return loadComponentInDiv({
    containerRequest: { url: "/fakeUrl/" },
    scope: fluidFileConverter.scope,
    urlResolver: new FakeUrlResolver(),
    documentServiceFactory: createLocalOdspDocumentServiceFactory(localOdspSnapshot),
    hostConfig
});

Comment (vladsud) [reply packages/tools/fluid-runner/src/exportFile.ts]: There will be no div - that's Node.js environment. Quickly glancing at the code, it references `window` object - this code should not compile as is (if Teja does proper work to ensure DOM types are not available in the code that compiles and packaged for this solution).

That's the list favorable option overall (irrespective of above) - your code should not care about all these details, and we should ideally not expose them.

Comment (leeviana) [reply packages/tools/fluid-runner/src/exportFile.ts]: Please ignore the div naming, that's old, note that no div is passed in or assumed in these inputs. Things that have to do with the window are similarly likely related to only sourceless iframe code that wouldn't be used here. We probably should look at the codeloader code though since I believe the script download mechanism it uses involves loading the code onto the window and reading it - though I'm not well versed in any codeloading mechanisms that don't involve this.

From the perspective of the details, these are details that would be in this code here (see the implementation of createContainerAndExecute). I'm not adding in any new parameters that are not already exposed in this code (and could be changed in this code)

Comment (leeviana) [reply packages/tools/fluid-runner/src/exportFile.ts]: My proposal is simply to replace your "createContainerAndExecute" call with the OWH "createContainerAndExecute" call. It is fundamentally exactly the same except it brings with it things like codeloader logic for all MSFT containers/JS containers and any nuanced/weird logical things needed like precisely placed awaits to handle fluid race condition bugs (and the removal of those awaits when those bugs can be fixed! etc.). These are all things you would need to manually discover and add here and we'd end up with two unnecessary copies of the same code/logic.

Comment (vladsud) [reply packages/tools/fluid-runner/src/exportFile.ts]: I do not want to expose FakeUrlResolver, createLocalOdspDocumentServiceFactory, and how file content is encoded (is it one buffer, or some data structure, or a set of files on disk) - all of that should be implementation details that are not leaked out, allowing us to change it (for example, move to FileDeltaStorageService if that makes sense). Another thing we can do - get rid of Loader altogether in this flow - it's Ok to create Container directly, removing one extra layer (fake driver factory) and hard-coded value (container URI). I really want to be able to do all these changes without any impact or slow down and the need to sync up.

I'd say it's Ok to provide functor like below to app specific function (reversing somewhat the flow), if it helps:
```ts
(scope, codeLoader) => { container: IContainer; logger: ITelemetryBaseLogger }
```

I think you refer (by awaits) to cases where old content is loaded, and you need to wait for latest state of document. This is not applicable here - we work with fully offline version of the document. If there are any bugs here, I'd prefer to learn about them and fix them.

Comment (leeviana) [reply packages/tools/fluid-runner/src/exportFile.ts]: I'm not proposing that we expose it any more than it's exposed in this file. We are referring to FakeUrlResolver here in this code and my code suggestion would live in the same place in the same file, not a different layer. From the perspective of the code consuming this file's capabilities, none of these details have to be exposed.

I don't think we really have a need to inject any app specific functionality into the loader, especially if there's no need for waitForOpCatchUp classes of problems (which I think is one of the last lingering ones we're working to get rid of - so it sounds like that fully offline version will always be caught up to the latest state?) All I'm suggesting is to reuse the OWH code which does exactly "new Loader" but with any shared telemetry/bugfixes etc.

There was also an option that I was working with someone on our team for introducing an optional Loader override in OWH's HostConfig, but it looks like that PR was abandoned - this is something it could make sense to do though, but again, my suggestion is for all of this code to live in the same layer so it shouldn't matter too much.

At the point in time when you're writing all the code to return a container, that is exactly what OWH does so you'd be replicating any shared hosting code needed at that layer and have no need for OWH since you have your own copy.

File: packages/tools/fluid-runner/src/exportFile.ts [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+    const logger = ChildLogger.create(fileLogger, "LocalSnapshotRunnerApp",
+        { all: { Event_Time: () => Date.now() } });
+
+    try {
+        await PerformanceEvent.timedExecAsync(logger, { eventName: "ExportFile" }, async () => {
+            // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+            const codeLoaderBundle = require(codeLoader);
+            if (!isCodeLoaderBundle(codeLoaderBundle)) {
+                const message = "Code loader bundle is not of type CodeLoaderBundle";
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/exportFile.ts:50]: I wonder if you want to return not just a string (in case of error), but some object (with event name, string, maybe some optional parameters) and caller is responsible for logging it and properly reporting to user. It will be less code overall to write and it will force consistency (for example, the very first return does not log anything to telemetry).

File: packages/tools/fluid-runner/src/exportFile.ts [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        { all: { Event_Time: () => Date.now() } });
+
+    try {
+        return await PerformanceEvent.timedExecAsync(logger, { eventName: "ExportFile" }, async () => {
+            // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
+            const codeLoaderBundle = require(codeLoader);
+            if (!isCodeLoaderBundle(codeLoaderBundle)) {
+                const eventName = clientArgsValidationError;
+                const message = "Code loader bundle is not of type CodeLoaderBundle";
+                logger.sendErrorEvent({
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/exportFile.ts:59]: I'd think the caller would log. Otherwise why add eventName to IExportFileResponse?

File: packages/tools/fluid-runner/src/exportFile.ts [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    fluidFileConverter: IFluidFileConverter,
+): Promise<Record<string, string>> {
+    const loader = new Loader({
+        urlResolver: new FakeUrlResolver(),
+        documentServiceFactory: createLocalOdspDocumentServiceFactory(localOdspSnapshot),
+        codeLoader: fluidFileConverter.codeLoader,
+        scope: fluidFileConverter.scope,
+    });
+
+    const container = await loader.resolve({ url: "/fakeUrl/" });
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/exportFile.ts:111]: We need to have IContainerLoadOptions.loadMode.opsBeforeReturn: "cached" to process any trailing ops in snapshot.
It looks like it's passed through     [LoaderHeader.loadMode]: IContainerLoadMode; header

File: packages/tools/fluid-runner/src/fakeUrlResolver.ts [hunk 20]
```diff
@@ -0,0 +1,49 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IRequest } from "@fluidframework/core-interfaces";
+import { IContainerPackageInfo, IResolvedUrl, IUrlResolver } from "@fluidframework/driver-definitions";
+import { IOdspResolvedUrl } from "@fluidframework/odsp-driver-definitions";
+
+const fakeId = "b9d30c31-dd2a-4658-bd01-cf77a7dec25c";
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/fakeUrlResolver.ts:10]: maybe use human readable string instead? Like fakeResolver? It's a bit easier for humans to realize that they can search for it in code base when they see it :)

File: packages/tools/fluid-runner/src/fluidRunner.ts [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                        describe: "Name of the file containing local ODSP snapshot",
+                        type: "string",
+                        demandOption: true,
+                    })
+                    .option("outputFolder", {
+                        describe: "Name of the output file",
+                        type: "string",
+                        demandOption: true,
+                    })
+                    .option("scenario", {
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/fluidRunner.ts:35]: At minimum, we need to pass this argument to createContainerAndExecute().
It would be better if it was more generic (like all the unknown args are passed to the createContainerAndExecute), such that solution is more versatile.

File: packages/tools/fluid-runner/src/logger/FileLogger.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    /** Hold events in memory until flushed */
+    private events: string[] = [];
+
+    /**
+     * @param filePath - file path to write logs to
+     * @param eventsPerFlush - number of events per flush
+     */
+    public constructor(
+        private readonly filePath: string,
+        private readonly eventsPerFlush: number,
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/logger/FileLogger.ts:24]: maybe have some default value?

File: packages/tools/fluid-runner/src/test/exportFile.spec.ts [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+import { MockLogger } from "@fluidframework/telemetry-utils";
+import { createContainerAndExecute, exportFile } from "../exportFile";
+import { fluidExport } from "./sampleCodeLoader";
+
+describe("exportFile", () => {
+    const folderRoot = path.join(__dirname, "../../src/test");
+    const outputFolder = path.join(folderRoot, "outputFolder");
+    const snapshotFolder = path.join(folderRoot, "localOdspSnapshots");
+
+    fs.readdirSync(snapshotFolder).forEach((snapshotFileName: string) => {
```
Comment (vladsud) [inline packages/tools/fluid-runner/src/test/exportFile.spec.ts:18]: I believe we do not validate that we are passing scenario through the stack

**Review by vladsud — APPROVED** [review]

File: packages/tools/fluid-runner/src/exportFile.ts (hunk 19 above)
Comment (vladsud) [reply packages/tools/fluid-runner/src/exportFile.ts:111]: And deltaConnection: "none" would be nice (but I think not required - due to usage of storageOnly mode)
